### ISSUE TITLE
Different result of cross entropy loss between result from model as function and using predict function

### ISSUE THREAD

**Issue by Xinshuai-Lyu** [keras-team/keras#16737]
**Describe the problem**.

Describe the problem clearly here. Be sure to convey here why it's a bug in Keras or why the requested feature is needed.

    Firstly, you get a prediction using model as a function, call this predition A; and then you use predict function of the model and get the same seemingly same prediction, call it B.

    If you use CategoricalCrossentropy to calculate loss based on same label and the same two results A and B, the loss are different between them.

**Describe the current behavior**.

    If you use CategoricalCrossentropy to calculate loss based on same label and the same two results A and B, the loss are different between them.

**Describe the expected behavior**.

    If you use CategoricalCrossentropy to calculate loss based on same label and the same two results A and B, the loss should be same.

**[Contributing](https://github.com/keras-team/keras/blob/master/CONTRIBUTING.md)**.

- Do you want to contribute a PR? (yes/no):
- If yes, please read [this page](https://github.com/keras-team/keras/blob/master/CONTRIBUTING.md) for instructions
- Briefly describe your candidate solution(if contributing):

**Standalone code to reproduce the issue**.

    class NLPModel(keras.Model):
      def __init__(self, **kwargs):
        super().__init__(**kwargs)
        self.hidden_layer1 = tf.keras.layers.GRU(400)
        self.outputs = tf.keras.layers.Dense(3, activation="sigmoid")
      def call(self, inputs):
        outputs = self.hidden_layer1(inputs)
        outputs = self.outputs(outputs)
        return outputs
    model = NLPModel()
    label = tf.constant([[1.,0.,0.]])
    prediction = model(tf.constant([[[0.4, 0.5, 0.54,0.4, 0.5, 0.54,0.4, 0.5, 0.54,0.9],
                                      [0.4, 0.5, 0.54,0.4, 0.5, 0.54,0.4, 0.5, 0.54,0.9]]]))
    cce = tf.keras.losses.CategoricalCrossentropy()
    loss = cce(label, prediction)
    def custom_cce(label, prediction):
      return tf.reduce_mean(tf.math.log(tf.matmul(label, tf.transpose(prediction)) / tf.reduce_sum(prediction)))
    print("before numpy")
    print(prediction)
    print(loss)
    print("before numpy")
    print("after custom_cce")
    print(custom_cce(label, prediction))
    print("after custom_cce")
    prediction = prediction.numpy()
    loss = cce(label, prediction)
    print("after numpy")
    print(prediction)
    print(loss)
    print("after numpy")

Provide a reproducible test case that is the bare minimum necessary to generate
the problem. If possible, please share a link to Colab/Jupyter/any notebook.

**Source code / logs**.

    before numpy
    tf.Tensor([[0.5021686  0.48505145 0.5026401 ]], shape=(1, 3), dtype=float32)
    tf.Tensor(1.0769439, shape=(), dtype=float32)
    before numpy
    after custom_cce
    tf.Tensor(-1.0875016, shape=(), dtype=float32)
    after custom_cce
    after numpy
    [[0.5021686  0.48505145 0.5026401 ]]
    tf.Tensor(1.0875016, shape=(), dtype=float32)
    after numpy

Include any logs or source code that would be helpful to diagnose the problem. If including tracebacks, please include the full traceback. Large logs and files should be attached. Try to provide a reproducible test case that is the bare minimum necessary to generate the problem.

**Comment by tilakrayal** [comment]
@gowthamkpr,
I was able to reproduce the issue on tensorflow v2.8, v2.9 and nightly. Kindly find the gist of it [here](https://colab.research.google.com/gist/tilakrayal/32ec62a6f1a747cf85fe2effa895b513/56555.ipynb).

**Comment by rchao** [comment]
This may be expected because the cross entropy operation is unstable and a suggestion would be to try with `float64` everywhere.
